test/test_portability.cpp

```diff
@@ -132,3 +132,58 @@ TEST_CASE("portableCopy works with all portability strategies", "[portableCopy]"
   // free device memory
   PORTABLE_FREE(a);
 }
+
+PORTABLE_FORCEINLINE_FUNCTION void expensive_operation(int iterations) {
+  double sum = 0.0;
+  for (int i = 0; i < iterations; i++) {
+    sum += 1.0 / (i + 1.0); // Expensive floating-point operation
+  }
+  // Prevent compiler optimization
+  if (sum < 0) {
+    printf("Artificial delay: %f\n", sum);
+  }
+}
+
+TEST_CASE("PORTABLE_FENCE properly synchronizes execution after a portableFor",
+          "[portableFor][PORTABLE_FENCE]") {
+
+  // Create host array
+  constexpr int N = 100;
+  constexpr Real init_val = 2.0;
+  constexpr Real mult_factor = 50.;
+  std::vector<Real> h_data(N); // Zero initialized
+  for (size_t i = 0; i < N; i++) {
+    h_data[i] = init_val;
+  }
+
+  // Copy to device
+  constexpr const size_t bytes = N * sizeof(Real);
+  Real *d_data = (Real *)PORTABLE_MALLOC(bytes);
+  portableCopyToDevice(d_data, h_data.data(), bytes);
+
+  // Use portableFor since it doesn't have an implicit fence
+  portableFor(
+      "Expensive loop to test PORTABLE_FENCE", 0, N, PORTABLE_LAMBDA(const int i) {
+        expensive_operation(1e6);
+        d_data[i] *= mult_factor;
+      });
+
+  // Fence before reduction
+  PORTABLE_FENCE("Fence after expensive operation");
+
+  // Use a device operation that would fail if the fence wasn't in place
+  int n_wrong = 0;
+  portableReduce(
+      "Check PORTABLE_FENCE", 0, N,
+      PORTABLE_LAMBDA(const int &i, int &n_wrong) {
+        if (d_data[i] != init_val * mult_factor) {
+          n_wrong += 1;
+        }
+      },
+      n_wrong);
+
+  REQUIRE(n_wrong == 0);
```

Comment:
I can't think of anything better, but I'm actually not sure this would necessarily fail if the fence wasn't in place... I *think* multiple `Kokkos::parallel_for`s in the same execution space somehow execute one-after-the-other. I have never fully understood the rules there.

Reply:
Yeaaaaaah I'm not sure. All I could discover was that there is an implicit fence at the end of a `parallel_reduce`, but that there _shouldn't_ be one at the end of `parallel_for`. I don't know what that means for what I've written though.

Reply:
Changed. I also increased the number of array items in case the first loop can continue on that task while the reduction goes in. The fence will ensure the answer doesn't change, but if somebody wants to test this, they can remove the fence and see what happens.